File: Data_Structure/Linked List/Doubly_Linked_LIst.cpp
// Doubli Linked List: Can access elements both the forword and backword direction
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node* prev;
};

void insert(Node*& head, int value) {
    Node* newNode = new Node{value, nullptr, nullptr};
    if(!head) {
        head = newNode;
    } else {
        Node* temp = head;
        while(temp -> next) {
            temp = temp -> next;
        }
        temp -> next = newNode;
        newNode -> prev = temp;
    }
}

void deletelist(Node*& head) {
    while(head) {
        Node* temp = head;
        head = head -> next;
        delete temp;
    }
}

void displayForword(Node* head) {
    while(head) {
        cout << head -> data << " <-> ";
        head = head -> next;
    }
    cout << "nullptr" << endl;
}

void displayBackword(Node* tail) {
    cout << "nullptr ";
    while(tail) {
        cout << tail -> data << " <-> ";
        tail = tail -> prev;
    }
    cout << endl;
}

int main() {

    Node* head = nullptr;
    insert(head, 1);
    insert(head, 2);
    insert(head, 3);
    
    Node* tail = head;
    while(tail -> next) {
        tail = tail -> next;
    }

    displayForword(head);
    displayBackword(tail);
    deletelist(head);

    return 0;
}

File: Data_Structure/Linked_List/Doubly_Linked_List.java
/*
Linked List: Doubly Linked List
Forward and backward direction 
*/

public class Doubly_Linked_List {

    private static class Node {
        int data;
        Node prev;
        Node next;

        Node(int data) {
            this.data = data;
            this.prev = null;
            this.next = null;
        }
    }

    public static void main(String[] args) {
        Node head = new Node(10);
        Node node1 = new Node(20);
        Node node2 = new Node(30);
        Node tail = new Node(40);
        
        head.next = node1;
        node1.next = node2;
        node2.next = tail;
        
        tail.prev = node2;
        node2.prev = node1;
        node1.prev = head;

        while (head != null) {
            System.out.print(head.data + " -> ");
            head = head.next;
        }
        System.out.println("null");

        while (tail != null) {
            System.out.print(tail.data + " -> ");
            tail = tail.prev;
        }
        System.out.println("null");
    }
}
